File: Flood_Fill_733.java
//Time Complexity : O(m*n)
//Space Complexity : O(m*n)
//Did this code successfully run on Leetcode : Yes
//Any problem you faced while coding this : None

package com.s30.satish;

import java.util.LinkedList;
import java.util.Queue;

//Use BFS or DFS approach to solve this problem as the time and space complexity will be O(m*n) for both approaches.
class Flood_Fill_733 {
    int color; // Initial color
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
    	// If the input matrix is empty or the image at the given location is newColor, then return image matrix
        if(image == null || image.length == 0 || image[sr][sc] == newColor)
            return image;
        // Store the initial color in the color variable
        color = image[sr][sc];
        // Define dirs array
        int[][] dirs = {{0, -1}, {-1, 0}, {0, 1}, {1, 0}};
        
        // Define Queue with type parameter as int[] to hold the row and column index of the matrix elements
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[] {sr, sc}); // Add the row and column index of the given location to start with
        image[sr][sc] = newColor; // Update the color with newColor i.e, perform flood filling
        
        while(!q.isEmpty())
        {
        	// Store the Queue element in temp array and delete from the Queue
            int[] temp = q.poll();
            
            // Iterate through all the neighbors of the current element of the image matrix
            for(int[] dir : dirs)
            {
                int row = temp[0] + dir[0];
                int col = temp[1] + dir[1];
                
             // Check the boundary conditions and check whether the element's color is initial color
                if(row >= 0 && col >= 0 && row < image.length && col < image[0].length && image[row][col] == color)
                {
                	// Update the color of the current element with the newColor
                    image[row][col] = newColor;
                    q.add(new int[] {row, col}); // Add the new location into the Queue
                }
            }
        }
        return image;
    }
}

File: floodfill733.py
"""
Approach 1 - BFS
TC - O(m*n) - O(V+E)- since traverses all elements in the matrix in the worst case
SC - O(m*n) - O(V) - for the queue
"""

class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        if (len(image) == 0 or image is None or image[sr][sc] == newColor):
            print("Returning")
            return image
        
        rows = len(image)
        columns = len(image[0])
        
        color = image[sr][sc]
        print("Color: ", color)
        image[sr][sc] = newColor
        
        queue = []
        queue. append(sr)
        queue.append(sc)
        
        directions = [(0,1), (0,-1), (1,0), (-1,0)]
        
        while queue:
            c = queue.pop()
            r = queue.pop()
            
            for direct in directions:
                dr = r + direct[0]
                dc = c + direct[1]
                
                if (dr >= 0 and dr < rows and dc >= 0 and dc < columns and image[dr][dc] == color):
                    image[dr][dc] = newColor
                    queue.append(dr)
                    queue.append(dc)
        
        return image
 
""" 
Approach 2 - DFS
TC - O(m*n)
SC - O(m*n) recursive stack space else O(1)

"""
class Solution:
    def __dfs(self, image: list[list[int]], sr: int, sc: int, newColor: int):
        
        # base
        if (sr < 0 or sc < 0 or sr == self.rows or sc == self.columns or image[sr][sc] != self.color):
            return
        
        # Setting the new Color
        image[sr][sc] = newColor
        # up, down, right, left
        directions = [(0,-1), (0,1), (1,0), (-1,0)]
        for direct in directions:
            r = sr + direct[0]
            c = sc + direct[1]
            self.__dfs(image, r, c, newColor)
                    
        
         
    
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        if (not image or len(image) == 0 or image[sr][sc] == newColor):
            return image
        
        self.rows = len(image)
        self.columns = len(image[0])
        
        self.color = image[sr][sc]
        
       # image[sr][sc] = newColor
        
        self.__dfs(image, sr, sc, newColor)
        return image
    
        
